### ISSUE TITLE
TypeError: require.cache.hasOwnProperty is not a function

### ISSUE THREAD

**Issue by wzup** [krakenjs/freshy#10]
This line in my code throws:

> freshy\index.js:54:23

Error:
```
  ● Test suite failed to run

    TypeError: require.cache.hasOwnProperty is not a function

      at unload (node_modules\freshy\index.js:54:23)
      at getFresh (node_modules\freshy\index.js:20:5)
      at node_modules\freshy\index.js:94:19
      at exports.create (node_modules\adaro\lib\engine.js:43:16)
      at Object.<anonymous> (app.js:53:26)
      at Object.<anonymous> (__tests__\one.test.js:8:11)
```

Why does it happen? How to solve?

**Comment by totherik** [comment]
`master` on Node v6.2.0:

```bash
$ npm test

> freshy@1.0.2 test /Users/ertoth/PayPal/src/git/freshy
> tape test/*.js

TAP version 13
# freshy
# reload
ok 1 should not be equal
ok 2 should not be equal
ok 3 should not be equal
ok 4 should be equal
ok 5 should be equal
# freshy
ok 6 should not be equal
ok 7 should be equal
# freshy with callback
loaded
ok 8 should be equal
ok 9 should not be equal
ok 10 should be equal
ok 11 should be equal
# unload
ok 12 (unnamed assert)
ok 13 should not be equal
ok 14 should not be equal
ok 15 should not be equal
ok 16 should be equal
ok 17 should be equal
# npm module
ok 18 should not be equal
ok 19 should be equal
# not yet loaded module
# abs path
ok 20 should not be equal
ok 21 should be equal
# node_modules require from deeply nested module
ok 22 should be equal

1..22
# tests 22
# pass  22

# ok
```

`master` on Node v7.7.3

```bash
$ npm test

> freshy@1.0.2 test /Users/ertoth/PayPal/src/git/freshy
> tape test/*.js

TAP version 13
# freshy
# reload
ok 1 should not be equal
ok 2 should not be equal
ok 3 should not be equal
ok 4 should be equal
ok 5 should be equal
# freshy
ok 6 should not be equal
ok 7 should be equal
# freshy with callback
loaded
ok 8 should be equal
ok 9 should not be equal
ok 10 should be equal
ok 11 should be equal
# unload
ok 12 (unnamed assert)
ok 13 should not be equal
ok 14 should not be equal
ok 15 should not be equal
ok 16 should be equal
ok 17 should be equal
# npm module
ok 18 should not be equal
ok 19 should be equal
# not yet loaded module
# abs path
ok 20 should not be equal
ok 21 should be equal
# node_modules require from deeply nested module
ok 22 should be equal

1..22
# tests 22
# pass  22

# ok
```

**Comment by wzup** [comment]
And what does it mean? I've just shown you an error that apparently is thrown by your plugin. Why do you show me your virtual test when real live code throws?

**Comment by totherik** [comment]
My apologies, when I saw `Test suite failed to run` I thought that was the test suite for this module.

Unfortunately, there's not enough context in this issue for me to debug or diagnose your problem. I don't have any context of your real live code and have no way to reproduce the issue. Usually, when filing issues, providing a pared-down, reproducible test case is the easiest way to share the scenario in which the problem is happening. So, if possible, please provide an isolated test case that exhibits this behavior.

**Comment by wzup** [comment]
@totherik,

I have a Node.js + React + Redux application.
I use [jest](https://facebook.github.io/jest/) as testing tool. A React's recommended test utility.

And when I set up my test environment and run a test, I get this error thrown
> TypeError: require.cache.hasOwnProperty is not a function

Here is my set up. Nothing special, I create an `ExpressJS` app and specify middlewares literally same as in my real app. But for some reason it works all right in a real app, and throws when testing. Why?

```
// Node utilities
const util = require('util');
const path = require('path');
const config = require('../config');
const request = require('supertest');
let jsdom = require("jsdom");

// ExpressJS
let express = require('express');
let routes = require('../routes/routes'); // ExpressJS routes and request handlers
const cookieParser = require('cookie-parser');
let bodyParser = require('body-parser');
var adaro = require('adaro'); // Here is Adaro that uses `freshy` plugin

// Create an app and specify middlewares
let app = express();
app.set('views', path.resolve(__dirname, 'views'));
app.engine('dust', adaro.dust({})); // !! Error starts here, it is "at Object.<anonymous> (__tests__\one.test.js:8:11)"
app.set('view engine', 'dust');

app.use(cookieParser());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({
    extended: false
}));
app.use(express.static(path.join(__dirname, 'public')));
app.use('/', routes);


// And a test itself
describe('Describe Test One', function() {
    it('Test One', (done) => {
        let agent = request.agent(app);
        agent
            .get('/')
            .set({'Cookie': ['sid=12345']})
            .withCredentials()
            .end((err, res) => {
                if(err) {
                    console.error('[ERROR]:', util.inspect(err, { depth: 1, colors: true }));
                    done();
                    return;
                }
                if(res.error) {
                    jsdom.env(res.error.text, (err, window) => {
                        console.log('[ERROR]:', window.document.documentElement.outerHTML);
                        done();
                        return;
                    })
                    return;
                }
                console.log('[response]:', util.inspect(res.body, { depth: 3, colors: true }));
                done();
            });
    });
});
```

So why does it happen? What do I miss?

Thank you

**Comment by totherik** [comment]
It looks like [Jest attempts to do a similar thing](https://github.com/facebook/jest/blob/27a807725811699ce817877e30afeb01ec9735b5/blog/2016-09-01-jest-15.md#module-registry-persistence) as freshy, but [replaces the cache with an object without a prototype](https://github.com/facebook/jest/blob/a2bab8d689ea53841ed8e7182f41e2c2ca45136f/packages/jest-runtime/src/index.js#L609) (meaning no Object methods, such as `hasOwnProperty`).

EDIT: It reads as though that behavior was removed by default and needs to be enabled, if desired. Either way, since this occurs under testing only I would investigate there to see if Jest might be doing something related to this in your setup.

**Comment by baldercm** [comment]
Same error here after upgrading from nodejs v7 to v8.

**EDIT:**
Easy to reproduce, the following will work on nodejs v7 but tests will fail on v8:
1. `git clone https://github.com/totherik/freshy`
2. `cd freshy`
3. `npm i`
4. `npm test`

**Comment by mrispoli24** [comment]
Hey guys,

This issue killed a  bunch of my projects I was able to fix it by going to the index.js file and changing line 54 to:

`if (Object.prototype.hasOwnProperty.call(require.cache, path) && require.cache[path].children) {...`

This got my projects running again.

**Comment by totherik** [comment]
Hey everyone, sorry about the delay. I'm away from a computer on vacation with my family. Thanks to #11 from @indutny, this _should_ be fixed. I just published 1.0.3 with his fix.

**Comment by Yazan-Ali-01** [comment]
> Hey guys,
> 
> This issue killed a bunch of my projects I was able to fix it by going to the index.js file and changing line 54 to:
> 
> `if (Object.prototype.hasOwnProperty.call(require.cache, path) && require.cache[path].children) {...`
> 
> This got my projects running again.

Thanks very much , That saved my life
